### ISSUE TITLE
Better phosphor effect

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#75]
We have been discussing for quite a while now. I think after we finished implementing the new core, we should come back to this.

**Comment by sa666666** [comment]
Yes, I have your patch here.  The only problem is that there are issues with PAL colour loss, which hasn't be implemented yet (https://github.com/stella-emu/stella/issues/13).

**Comment by thrust26** [comment]
I have seen quite some game development lately which seems to rely on Stella's very forgiving phosphor emulation too much. So maybe we should add more focus to this.

**Comment by ale-79** [comment]
Related to this, currently the phosphor effect is applied also to the TIA display in the debugger (but not on the "zoom" area). This is not a bug, as it's explicitly mentioned in the documentation
Anyway, IMO in the debugger you should only see the raw TIA output, without any filter applied.
What do you think?

**Comment by thrust26** [comment]
I would prefer raw output here too. But its no big deal, because we have the zoom window.

**Comment by sa666666** [comment]
This is now committed in https://github.com/stella-emu/stella/commit/f8ea61875c5ea379b798baf3322be1d9b2c2b13f.  Currently it's not working in Blargg mode; this is the next TODO item.

**Comment by thrust26** [comment]
Probably we should use much lower default blending values (and convert existing ones). Something around 25 (~1/3rd of 77) seems more realistic.

**Comment by sa666666** [comment]
The current implementation looks good at 77.  And unless I'm misunderstanding the code, anything below 50 is as if phosphor mode isn't enabled at all (which is why I've capped the lower limit to 50).  I think we need another pre-release to experiment with this.  I will look into it once I get phosphor mode and Blargg working again.

**Comment by thrust26** [comment]
@sa666666: Not sure why you assume anything below 50 disables phosphor. If e.g. the current value is 0 and the previous value was 100, getPhosphor will return 100 * (blend value/100). So a blend value of 25 would return 25 here.

Unfortunately I cannot really test lower values, because my LCD is **very** slow. So I tried with the debugger, but (unlike the 4.7.x versions, where it is visible in the TIA Display, but not in the TIA Zoom), there the phosphor effect is not displayed at all.

High blend values close to 100 result into some very cool effects. :)
![asteroids 1981 atari _5](https://cloud.githubusercontent.com/assets/24360774/26542997/62285ae6-445c-11e7-8654-1b48846a12a1.png)

BTW: Changing the values and enabling/disabling the phosphor mode is not immediate. You either have to reload the ROM or enter and leave the debugger. Immediate would be nice, else the menu needs a (*).

**Comment by sa666666** [comment]
My mistake on the assumption of below 50 having no effect; it looks like I was using some test code when I made that claim.  So I've changed the code to have 0-100 as the valid range.  I will get the binaries updated later this evening to reflect that.

The phosphor effect is no longer displayed in the debugger, since with this new code the effect is spread out over time, and the debugger shows only one frame at a time.  Previously, the phosphor frame was blended before being shown, so the entire frame was visible.  Unfortunately this will also have consequences for taking snapshots in that mode.  In any event, a request further up in this issue indicated that you guys would rather that the debugger showed raw output anyway.

As for changing the phosphor mode within the Game Info dialog, it was never immediate in that case.  There is already a '*' at the bottom of the dialog indicating as such.  Now, if you use the shortcut key (Alt-p), then phosphor mode is toggle immediately.  What isn't accessible immediately is the blend amount.  It would be easy enough to add a key for the next release.

**Comment by thrust26** [comment]
> In any event, a request further up in this issue indicated that you guys would rather that the debugger showed raw output anyway.

Yup, and that's fine.

BTW: I noticed that the default in the code is 77 while in the menu it is always 60. How that?

> As for changing the phosphor mode within the Game Info dialog, it was never immediate in that case. There is already a '*' at the bottom of the dialog indicating as such.

Ah, my bad. I always thought this would be a footnote, referring to some entries marked with (*).

**Comment by sa666666** [comment]
Another error on my part.  For the next pre-release (yes, there will be one :smile:), the level is changed to 30 everywhere, and the range is set to 0 - 100.  And there will be shortcut keys to change phosphor blend when actually playing a game (it will be immediate).

**Comment by thrust26** [comment]
Doesn't Stella have defaults for certain games? If yes, those should be adjusted (by 30/77, based on the new default).

**Comment by sa666666** [comment]
There are only 3 on the properties database that use something other than the default (66, I believe).  I will look at these.  Also, I may have to adjust some new ones, particularly Compumate, since it's pretty unreadable at 30.

**Comment by thrust26** [comment]
Do you have VSync enabled? For me that makes quite a difference.

**Comment by sa666666** [comment]
I always have vsync enabled.  To me, it is/should be a requirement.

I've just committed some fixes in https://github.com/stella-emu/stella/commit/5dbd9fee5103aa31b1a1ff2db958bc567e3a3765.  The range of blending is now 0 - 100, the default is 30, and the Alt-i and Alt-o keys decrease/increase blending dynamically.  I will do another pre-release when I get home this afternoon.

**Comment by SpiceWare** [comment]
Looks like the decay is still occurring when you enter command mode by hitting \, which is how I usually take snapshots (due to the Mac's repurposing of the function keys).

Screenshot using F12 - can see decay within the station in the title:
![draconian](https://cloud.githubusercontent.com/assets/26531922/26553199/3a18678c-4450-11e7-8cba-c2f0312ec1f3.png)

Screenshot using \ then clicking the snapshot button - no decay within the title
![draconian_1](https://cloud.githubusercontent.com/assets/26531922/26553202/4079175c-4450-11e7-9d9e-e902829029cb.png)

If you set decay to 99 you can clearly see the decay still happening if you hit \ and leave the menu up.

Decay at 99 works well for screenshots of static interlaced displays, you can subtly see the flicker:
![draconian_3](https://cloud.githubusercontent.com/assets/26531922/26553338/c168ba98-4450-11e7-90fa-8a9efdaab314.png)

Decay at 100 not so much 😆 
![draconian_2](https://cloud.githubusercontent.com/assets/26531922/26553357/dde54380-4450-11e7-9ff7-ae5354f50b00.png)

**Comment by thrust26** [comment]
I wonder if we should rename that value. Is Blend still the correct term here?

**Comment by sa666666** [comment]
I leave that to you guys to decide.  To me, it's just a string that can be easily replaced with something else.

**Comment by thrust26** [comment]
I don't care, just wondering.

**Comment by thrust26** [comment]
Something in the code seems wrong. 

With a blend value of 50, when switching between black and white, the black frame should be displayed with 50% (118) of the white RGB values (236). But the RGB values of the attached test ROM are just 1/4 (59). 
With blend = 66, the expected values is 2/3 or 236 (~156), but the result is 4/9 (102).

So somehow the values are blended twice. Where is the error?
[test.zip](https://github.com/stella-emu/stella/files/1036602/test.zip)

BTW: Is it intentional that the values created by ALT+I/O are not stored in the properties?

**Comment by sa666666** [comment]
How are you determining this math; what are the equations used?  It's possible that something is wrong in the code, but this is the code you sent me.  Or maybe I modified something incorrectly??

As for you second point, yes, changing ROM properties dynamically will not store the ROM properties.  This is also true for toggling phosphor mode, format, color/bw, etc.  If you want to save the properties, simply go to 'Game Properties' and click 'OK'.  At that point they are saved.  The reasoning was that changing things dynamically was meant to show how things would look when changed, and then when you want to keep the changes, you have to manually do so.

I will look at the phosphor code again tomorrow morning to see what's going on.

**Comment by thrust26** [comment]
I simply made some screenshots from the test ROM and checked the RGB values (all the same for grays).

**Comment by sa666666** [comment]
TIASurface::pixel() is a convenience function only used in the debugger; TIASurface::render() is the function used by during emulation mode to render the frame.  At no point are these two called together.

I'm looking at the screenshot code now, and don't see any issues.  I will explore this further, and compare its output with ones generated by the OS itself.

**Comment by thrust26** [comment]
Here are two screenshots from Asteroids (Blend = 50). The asteroids are flickering at 30Hz. Notice how much darker they are in Stella.
Stella:
![asteroids 1981 atari _7](https://cloud.githubusercontent.com/assets/24360774/26763054/89347152-494c-11e7-9c23-5e210a3e144e.png)
OS:
![asteroids_os](https://cloud.githubusercontent.com/assets/24360774/26763064/a76aa3ee-494c-11e7-964a-eadee0369b2f.png)

**Comment by sa666666** [comment]
I am still having major issues with this.  On my system, a default blend level of 30 is basically the same as not having phosphor effect turned on at all; it flickers just as much as.  For me, I need to increase it to 80 or so before the flicker starts to disappear, but is still there enough to simulate the real thing.  That's why I'm also implementing #144.

Also, I can't see how to fix the issue of a snapshot taking a darkened image.  The relevant code takes a snapshot of the current frame, and whatever is on the screen at that time is going to be captured.  And half of the time, that image will be slightly darkened.  This is really a symptom of a larger problem; how do you take a discrete snapshot of a continuous event (multiple frames)??

**Comment by thrust26** [comment]
I just press F12 multiple times. With 30Hz flicker there is a 50% chance of hitting the right frame.

**Comment by sa666666** [comment]
Here are several PNGs generated by pressing F12 fairly quickly in succession.  Note that it's doing exactly what you'd expect.  In some frames the ship is visible and the asteroids are dark; in others it's vice-versa.  Note that without phosphor mode enabled, the dark images would disappear altogether.  But it is still only taking a snapshot of a single screen at a time.  The fact that one doesn't see that in emulation is an optical illusion.  I guess the same would happen on a real TV.

[asteroids.zip](https://github.com/stella-emu/stella/files/1066652/asteroids.zip)

**Comment by thrust26** [comment]
Can you post your screenshots? Stella and OS when the asteroids are dark.

**Comment by sa666666** [comment]
The screenshots are in the attached ZIP file :)  Taking them from Stella or the OS (KDE, LInux) doesn't matter.  Here are a few:
![asteroids 1981 atari no copyright](https://user-images.githubusercontent.com/16061066/27013551-8f611a0c-4ec0-11e7-87b3-d52b534e574d.png)
![asteroids 1981 atari no copyright _1](https://user-images.githubusercontent.com/16061066/27013553-93874e4e-4ec0-11e7-8dd5-4d35ab0c7eb7.png)

**Comment by thrust26** [comment]
I compared the RGBs of the yellow asteroids:

Yours (Stella, correct?):
Enabled: 187, 187, 53
Dark 1: 46, 46, 13
Dark 2: 23, 23, 6 (from the trace of the enabled asteroids)
Dark 3: 11, 11, 3 (from the trace of the Dark 1 asteroids)

Mine (Stella):
Enabled: 187, 187, 53
Dark 1: 46, 46, 13
Dark 2: 23, 23, 6
Dark 3: 11, 11, 3

Mine (OS):
Enabled: 187, 187, 53
Dark 1: 93, 93, 26
Dark 2: 46, 46, 13
Dark 2: 23, 23, 6

Only my OS screenshots (can you post some too?) are like expected. And IMO they should speak the truth, because one error path (screenshots) is eliminated. If you look at the values, for me it seems very clear, that when taking a screenshot the disabled objects are blended once more. 

I think this happens in TIASurface::pixel. This is called by TIASurface::baseSurface which is only called by EventHandler::takeSnapshot.

**Comment by sa666666** [comment]
When taking snapshots and not in 'ss1x' mode (ie, snapshots are always 2x or bigger, all effects enabled, etc), then TIASurface::pixel is not being called.  I verified this by adding a print statement to the methods you mention above; they are not being called.  So they cannot be causing this issue.

Here are some snapshots from the OS itself:
![screenshot_20170611_165546](https://user-images.githubusercontent.com/16061066/27013883-ea557a24-4ec6-11e7-8cd2-bcf7693aa0e3.png)
![screenshot_20170611_165547](https://user-images.githubusercontent.com/16061066/27013882-ea5354b0-4ec6-11e7-8c4d-5806d598294f.png)

**Comment by DirtyHairy** [comment]
Hmmm, I just checked this out, and I can add two observations:

 * I have to switch phosphor off and back on for any change of the blend settings to take effect
 * For me, a value of 60 - 70 is required to get a meaningful reduction of flicker

I have a feeling that the differences between the amount of flicker reduction that we see is due to different postprocessing in our video stack --- OS, video driver and display. If these boost low intensities, then lower blend levels should be sufficient.

**Comment by sa666666** [comment]
I have to use 60-70 as well, but I believe we're both using Linux, so maybe that is the commonality.  I don't have to toggle phosphor mode to see a difference, though; using Alt-i and Alt-o to change the blend level immediately starts to show a difference.

This is also why I think we need #144 implemented; there will never be a hardcoded default value that works for everyone.

**Comment by DirtyHairy** [comment]
>  I don't have to toggle phosphor mode to see a difference, though; using Alt-i and Alt-o to change the blend level immediately starts to show a difference.

And there we have the difference: I was using the menu :stuck_out_tongue

> This is also why I think we need #144 implemented; there will never be a hardcoded default value that works for everyone.

:+1: I agree.

**Comment by DirtyHairy** [comment]
It just occurred to me: @sa666666, what video driver are you using on linux? For me, it's an intel IGP with the usual opensource drivers.

**Comment by sa666666** [comment]
Update: the post above with the OS snapshots have been updated; I was using 30% blend, when the others posted before it were 50%.  These new ones are 50% too.

**Comment by sa666666** [comment]
AMDgpu drivers, open source (with an RX 480).

**Comment by thrust26** [comment]
For me 30 works nicely. Probably you are right and this is OS related. Still weird.

And Stephens OS screenshot provide the same values as my OS screenshots. Now, why are the disabled objects in Stella's screenshots darker by one blend factor?

**Comment by sa666666** [comment]
Confirmed that Stella-generated screenshots are applying one more blend level.  This must be in Stella, since we now have OS screenshots from Windows and Linux which are the same.

**Comment by thrust26** [comment]
And after thinking about it, it **MUST** be the different LCDs we are using (and not the OS). The OS screenshots provide the exact same values for us.

Probably the grey to grey fall time of my display is pretty long.

**Comment by sa666666** [comment]
I've found the problem with Stella applying blending one too many times.  A past issue was that snapshots would sometimes have UI messages erroneously overlaid on the TIA image, so I added code that essentially turned off the messages and rendered the frame once, to erase everything.  Turns out this is interfering with correct snapshot output.  When I disable lines 1859-1860 in EventHandler.cxx, Stella generates snapshots as follows:
![asteroids 1981 atari no copyright _3](https://user-images.githubusercontent.com/16061066/27014127-baac8672-4ecc-11e7-8185-360129fa0d52.png)
![asteroids 1981 atari no copyright _4](https://user-images.githubusercontent.com/16061066/27014126-baabb88c-4ecc-11e7-84c6-1e37c8589dcd.png)

So we now see the correct output, but the messages aren't being removed.  I guess I will have to figure out how to deal with that another way.

However, we still have the issue of blend level 30 being fine on @thrust26 system, and way too low on our systems.  This can't be taken care of with a set value, so we need to implement #144.  Problem is, what should the default value be, keeping in mind that many people never change the defaults??

**Comment by thrust26** [comment]
See above, use your values. My LCD is most likely not representative at all.

**Comment by DirtyHairy** [comment]
> And after thinking about it, it MUST be the different LCDs we are using (and not the OS). The OS screenshots provide the exact same values for us.

This is likely, allthough I am not completely convinced :smirk: I think it is possible that there is some postprocessing (e.g. gamma correction) happening above the layer in which the screenshot is captured. Otherwise, the screenshot would look different from the screen on the same machine due to double postprocessing.

Anyway, I think that augmenting the documentation is the only the viable option. This is the same thing as gamma correction in games: everyone has to configure it for themselves.

**Comment by thrust26** [comment]
I have never heard about any software post processing above the screenshot level. At least not in Windows. Do you have any links? 

What monitor are you using? And how is it set up (e.g. frequency and acceleration level)? Maybe we can identify, why your and @sa666666's monitor(s) need more blending than mine. So far my best guess is the response time.

**Comment by DirtyHairy** [comment]
> I have never heard about any software post processing above the screenshot level. At least not in Windows. Do you have any links?

None, and I am not sure whether such a thing exists, I am just trying to guess what might be going on :smirk: What I could image is that some settings like color temperature (I remember the NVIDIA driver has such options) or ICC profiles don't take effect in screenshots. This would at least make sense to me as such settings adjust how the screen content is presented on the display, while a screenshot captures the screen contents.

Regarding display: phosphor behaves pretty much the same on my laptop display and an external Dell LCD (displayport, both 60Hz refresh, I think). I just checked what happens if I turn up gamma using xrandr, flicker gets a bit better, but 30 still is too low.

**Comment by thrust26** [comment]
Don't forget that phosphor should only **reduce** flicker (to match TV flicker), not completely eliminate it like the old one.

**Comment by DirtyHairy** [comment]
:smirk: I know. To have a baseline, cranking blend up to 80 gives a flicker that is comparable to 6502.ts in WebGL mode (I am using a shader for flicker reduction there, so it doesn't work with WebGL off), although the asteroids have slightly longer trails in Stella.

**Comment by sa666666** [comment]
In Windows, about 50 is just right (a slight amount of flicker, but not too much).  In Linux, I also need to go to about 80 for the same effect.  I haven't tested in OSX yet.  Note that the Linux and Windows tests are on different systems, but the OSX one will be on the same system as Windows.

Right now I'm leaning towards making 50 the default, and let users adjust from there.

**Comment by DirtyHairy** [comment]
I think that this is a good way to proceed, in particular as Windows likely applies to the overwhelming majority of users.

**Comment by sa666666** [comment]
OK, 50 it is then.  I hope to get this finished by the end of the week, now that work is slowing down a little (midterm break).  After that is #144, and then I'm done I think.

@DirtyHairy, could you look at the latest issues and determine whether they can be done for 5.0 or should be pushed to 5.1 or so?

**Comment by DirtyHairy** [comment]
Ups, thanks for the pointer, #150 and #149 completely slipped me --- life is pretty busy atm. I'll take a look and comment there.

**Comment by thrust26** [comment]
50 is fine by me too. Plus some words in docs.

**Comment by sa666666** [comment]
So I'm still working on this, and trying to find a way to make things as fast as possible.  First some info on how everything works:

The current Blargg code works by first telling it about your system palette (in our case, the 256 colour TIA palette), and then it pre-computes another internal palette based on that, with various filtering applied).  Then when rendering happens, it converts every two TIA pixels into 7 Blargg pixels.  So during rendering, there are only 14 array accesses per pixel, in a framebuffer of approx. 160 x 210 or so (height is variable obviously, up to 320).  So something like the following:

// algorithm ///////////////////////////////////////////
pass palette to Blargg code  // done once
Blargg code creates its own palette  // done once

// following done every frame
for every 2 pixels in TIA:  convert to 7 pixels in Blargg
return buffer
///////////////////////////////////////////////////////////////

The above works fine when the palette is set and unchanging, as it is for normal, non-phosphor rendering.  Now we want to integrate the new phosphor code.  Problem is, this one is essentially generating a new palette each frame, which then needs to be passed to the Blargg code, it then creates its own palette, and then 14 array lookups happen per pixel.  The problem is the algorithm now becomes:

// algorithm ///////////////////////////////////////////
pass palette to Blargg code  // done each frame
Blargg code creates its own palette  // done each frame
for every 2 pixels in TIA:  convert to 7 pixels in Blargg  // done each frame
return buffer
///////////////////////////////////////////////////////////////

The Blargg code seems to be written so that all the intensive work is done in its palette creation code, and it's meant to be done only once.  That leaves a simple 14-point lookup for every 2 pixels.  Now, much more work is being done per frame, since the palette is constantly changing.

I haven't implemented this yet, since I'm trying to find a faster way of doing things.  I guess I should just try it and see how slow it will be.  But it would be nice to have a better design.

Comments, questions, etc?

**Comment by DirtyHairy** [comment]
I see a fundamental issue here: the phosphor code operates in RGB color space, and the result generally won't map to a discrete palette at all. We'd have to use a quantization algorithm to reduce it to 256 colors, and that would be dead slow.

I haven't looked at the Blargg code yet, but I think a viable approach would be:

* Replace the palette lookups in Blargg with function calls
* Create a Blargg palette from the TIA palette, store the indices in a lookup table (std::map should do) and try to map the 24 bit colors to precomputed palette values
* Perform the necessary calculations on the fly if the lookup fails

Most parts of a typical game screen are static and should not differ from the TIA palette, so the overhead will only apply to the moving parts. In addition, this algorithm should be only marginally slower than Stella 4 if phosphor is turned off (provided the calculation is actually slower than the lookup if potential cache misses are taken into account).

If this fails, we could also apply phosphor *after* Blargg --- no changes would be required for this approach at all :smirk:

**Comment by thrust26** [comment]
The Blargg palette issue was a major problem for me too, when I created the phosphor code (though I didn't fully understand it back then). 

A game like Asteroids (up to 8 big asteroids ones IIRC) might be a test case for slow calculations, as it flickers the asteroids. 

Phosphor mixes the previous **displayed** frame (which would be after Blargg, correct?) with the currently **rendered** frame. Then we should use phosphor after Blargg for the latter too. So the Blargg code wouldn't have to be changed, but phosphor would become a post processing step then. (if we would do phosphor before Blargg, we would mix a "Blargged" frame with a non-"Blargged" one, which makes no sense).

I am not sure if phosphor after Blargg will work very well, but it could be a good workaround for 5.0. Maybe a mixed solution would be most correct (some effects happen before phosphor and some after it). I am no expert here though and we can discuss this later. 

BTW: The TIA palette has only 128 entries (for NTSC). And aren't that only 3.5 array accesses per pixel?

**Comment by sa666666** [comment]
Yes, that's what I was basically trying to say.  Currently, there are only ever 256 RGB colours, so they can be added to an array-based lookup table (aka, a palette).  The RGB data created from the phosphor code changes per frame, and while there will only be 256 of them at a time, the actual 256 will change per frame.

I like the idea of using a map to store RGB -> Blargg conversions.  However, it's not just the RGB data that needs to be cached.  There is also a 'kernel' that is created based on the RGB values.  If you cache one, you need to cache the other too.

**Comment by DirtyHairy** [comment]
In order for phosphor before Blargg to work, we'll have to retain the frame _before_ applying Blargg and have phosphor mix with it --- mixing with the "Blargged" one makes no sense indeed.

**Comment by DirtyHairy** [comment]
>  Currently, there are only ever 256 RGB colours, so they can be added to an array-based lookup table (aka, a palette).

Interesting. I guess I'll just have to look at the code tonight, but how does this work? If I understand correctly, the phosphor code calculates a weighted average between its two input frames which can be anywhere in RGB space. How does the quantization work?

> There is also a 'kernel' that is created based on the RGB values

Ouch. If that depends on the whole palette, then caching lookups won't be of much help as each frame will esentially create a new cache entry.

**Comment by sa666666** [comment]
We always have access to the previous, pre-Blargg frame.  Some confusion here, so clarification is in order:

TIA::myFramebuffer -> holds current TIA frame in a 256 entry array-based, RGB palette.

TIASurface::myRGBFramebuffer -> holds combination of current RGB frame mixed with TIA frame (no Blargg applied yet).  This is what you call the 'currently rendered frame'.  The previously displayed frame also comes from this, before Blargg is applied.  Remember, Blargg and phosphor only interact when both are enabled. 

What is output by Blargg -> whatever RGB data is passed to Blargg code is then processed and output into the TIASurface buffer itself.  TIASurface contains a 'FBSurface', which has its own pointer to pixel data (it is essentially the final place for data to go, to be presented to the screen).

Also, while there are 128 colours in the TIA palette, we also have 128 for black and white.  So 256 total.  We need those other 128 colours to quickly process PAL colour loss.  Also, they're used when greying out the TIA image in the debugger.

**Comment by thrust26** [comment]
I am for applying phosphor by mixing the previous after Blargg and phosphor frame with the current frame after Blargg.  So Blargg can still rely on its own, pre-computed palettes.
That should make it as fast as now and relatively easy, doesn't it?

**Comment by DirtyHairy** [comment]
Considering the trouble with the kernels, I agree: let's apply phosphor *after* Blargg.

**Comment by sa666666** [comment]
@thrust26, no it won't be faster, for several reasons:

- you're operating on a larger surface now; 560x210 now instead of 160x210, since every 2 pixels have been converted to 7 pixels, so the horizonal resolution is now converted from 160 to 560
- the scanlines have been applied; ideally, the scanlines should be applied after everything, since they're an artifact of the TV screen itself, not NTSC effect

**Comment by DirtyHairy** [comment]
@sa666666 I just checked the code; to me it seems that `TIASurface::getRGBPhosphor` actually calculates a full 24 bit RGB triple which is then passed to SDL for conversion to the platform format (which is requested as 8bit RGBA). If I am not mistaken, then there are no palettized values for Blargg to start with.

**Comment by sa666666** [comment]
There is another way:

- compute TIASurface::myRGBFramebuffer each frame, as we do now
- tell the Blargg code about it each frame (which implies re-calculating all the palette/kernel entries
- pass TIA::myFramebuffer to Blargg code, and use its palette indices to render the frame

I think this will work, since if the TIA, for example, wants to use colour index 7, that 7 will correspond to raw colour 7 in the TIA palette, and RGB colour 7 in the RGB palette.  Furthermore, it will correspond to Blargg colour 7 in the Blargg palette. IOW, they're all essentially parallel arrays with corresponding RGB data.

@DirtyHairy, maybe I'm not using the word palette correctly.  The entire rendering system using 32 ARGB data.  The only thing that uses a palette (aka, an array-based lookup table) is the TIA itself.  And those indices map to 32-bit colours.

So we're back to the same issue I had at the beginning.  How do we adapt the current code to work with a frame that has a constantly changing 256-colour RGB palette?  I guess we just re-compute the palette every frame.

Another alternative is to say that phosphor and Blargg don't work together, and release it for 5.0 as-is.

**Comment by thrust26** [comment]
@sa666666 
- I see. But the larger surface is only relevant for the rather simple phosphor calculation, correct. So Blargg would still be as fast as today.
- If I understand scanlines correct, they are resulting from the beam movement. Which also happens before phosphor comes into play. 

For a progressive display (which we have here) in theory things would get more complicated, but as far as I can tell, that's not what Blargg currently emulates. 

**IF** we want to be more faithful here, we would have to created a surface with a double vertical resolution where the scanlines flickers interlaced at 30 Hz (can Blargg create this?). And then add phosphor.

**Comment by thrust26** [comment]
> So we're back to the same issue I had at the beginning. How do we adapt the current code to work with a frame that has a constantly changing 256-colour RGB palette? I guess we just re-compute the palette every frame.
If we apply phosphor **after** Blargg, then the palette isn't changing. Or do I miss something here?

**Comment by sa666666** [comment]
If we apply phosphor after Blargg, the colours have still changed; phosphor mode creates colours that didn't exist in the base TIA palette.  It does this by definition.

**Comment by thrust26** [comment]
Call me stupid, but I don't get you. :)

Blargg will only see the TIA palette, create new RGB values which will then be processed by phosphor, no?

**Comment by sa666666** [comment]
@thrust26 

Say we have only a 4 colour palette. So the TIA has only 4 colours (and also pretend they are only 8-bit colours), say [0x10, 0x20, 0x30, 0x40].

The Blargg code has a method called initializePalette().  It takes a pointer to theese 4 colours, and processes them in some fashion, generating new colours that it stores internally.

Now, phosphor mode is applied, and all 4 colours have changed.  So we need to call initializePalette() again, so that the Blargg code knows about the new colours, and in turn generates new internal colours.  Since phosphor mode potentially changes the colours every frame, initializePalette() needs to be called every frame.

Whether phosphor mode is applied before or after Blargg is irrelevant; the colours currently on the screen will be potentially different every frame, and initializePalette() therefore needs to be called every frame.

My initial question was that if there's a way to not call initializePalette() each frame.  But there doesn't appear to be.

**Comment by thrust26** [comment]
@sa666666 
Agreed with the first two paragraphs. 

But then, why do we have to call initializePalette() again? Why does Blargg have to know about colors changed by phosphor or what color are displayed on screen?

**Comment by DirtyHairy** [comment]
@sa666666 

I don't understand either :smirk: If we leave out phosphor, then Blargg just sees the 256 colors in the TIA palette, and those don't change. If we apply phosphor to the frames **after** they have been processed by Blargg, then nothing changes as far as Blargg is concerned: it still sees only the TIA palette. Where is the error in my thinking?

**Comment by sa666666** [comment]
Maybe I am confused now, since two people agree and one doesn't :smile: 

I think what I'm not stating clearly is that the colours passed to Blargg aren't the ones rendered to the screen.  Blargg takes those colours and applies filters to them, and converts them to YIQ format.  And those YIQ colours are what is placed in the output framebuffer.  So if phosphor mode changes the initial colours, then Blargg **must** see them, to calculate its own new colours.

**Comment by thrust26** [comment]
Where does phosphor change the initial colors? It comes after Blargg.

Phosphor has to work on the YIQ colors (as I just learned) created by Blargg, **not** vice versa. And then the result from phosphor is placed into the final framebuffer.

**Comment by sa666666** [comment]
I think I finally got it :bulb: I was still thinking about doing phosphor before Blargg, not after.  OK, let's try that that approach.

BTW, I just tried my suggestion (changing the palette before calling phosphor), and it's not working.  So best if we can just not have to touch the Blargg code at all.

**Comment by thrust26** [comment]
Puh, in between I almost started to question my sanity. 😄

**Comment by sa666666** [comment]
After 75 comments, we finally finish with issue #75 (is that a coincidence or what?).  Implemented in https://github.com/stella-emu/stella/commit/09af9d09a5947f6eacb999262c68ed05c94a41fe.

**Comment by thrust26** [comment]
That went fast. Does the result look ok?

**Comment by sa666666** [comment]
Result looks fine to me.  I will be posting a pre-10 release tonight.

It ended up not being a difficult problem once we hashed it out back and forth.  In particular, I never considered the case of blending after Blargg, so it wasn't even in my mind.  Even when you guys pointed it out, it took some time for it to sink in :imp: 

Sometimes just talking about an approach helps figure things out, even before coding starts.

**Comment by DirtyHairy** [comment]
> Sometimes just talking about an approach helps figure things out, even before coding starts.

If you have a hammer, everything looks like a nail I suppose :smirk:

**Comment by thrust26** [comment]
@sa666666 @DirtyHairy Some cool finding about how slow my LCD is: 

The attached ROM looks quite odd on my slow LCD. **Without** phosphor, the black smiling face shows some heavy trails when moving, **with** phosphor enabled (blend = 50) it looks much sharper. I would have expected the opposite. So at first glance I thought the modes were reversed and was about to open an issue. But then I realized that it must be my LCD, which doesn't like switching from black to something only slightly brighter. 

I suppose the effect is not visible on your LCDs?

[Hacker_Face.zip](https://github.com/stella-emu/stella/files/1172527/Hacker_Face.zip)

**Comment by DirtyHairy** [comment]
That's interesting. I can see the same effect, although much less pronounced than what you describe: on my laptop screen, I get some slight trails when moving, and enabling phosphor removes them :smirk:
